adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2020 Petuum, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import math
+import numpy as np
+import warnings
+
+from types import MethodType
+
+from adaptdl.torch.data import current_dataloader
+
+
+__all__ = ["ScalingRuleBase", "AdaScale", "LinearScale", "SqrtScale",
+           "LEGWScale"]
+
+
+class ScalingRuleBase(object):
+    """
+    Base class for scaling rules that has the ability to track gradient noise
+    scale calculations. Its subclasses can be used in combination with
+    ``adaptdl.torch.parallel.AdaptiveDataParallel`` and ``torch.optim.SGD``.
+
+    .. code-block:: python
+
+        optim = torch.optim.SGD(model, lr=0.001)
+        adascale = AdaScale()
+        model = AdaptiveDataParallel(model, optim, adascale)
+
+        for epoch in ...:
+            for batch in ...:
+                optim.zero_grad()
+                loss = ...
+                loss.backward()
+                adascale.step()
+    """
+    def __init__(self):
+        # instance of AdaptiveDataParallel, needs to be set before any of the
+        # methods can be used
+        self.adp = None
+        self._optimizer = None
+        self._orig_optimizer_step = None
+
+    def scale_lr(self, scale):
+        raise NotImplementedError
+
+    def zero_grad(self, *args, **kwargs):
+        if self.adp.gns._should_zero_grad:
+            self.adp.gns._reset_accumulation(*args, **kwargs)
```

Comment:
`_reset_accumulation` should be made public, and `_should_zero_grad` accessed through a `@property def should_zero_grad`